src/Domain/CoordinateMaps/TimeDependent/Shape.cpp

```diff
@@ -0,0 +1,368 @@
+  auto theta_phis = cartesian_to_spherical(centered_coords);
+
+  // The Cartesian gradient cannot be represented exactly by `l_max_` and
+  // `m_max_` which causes an aliasing error. We need an additional order to
+  // represent it. This is in theory not needed for the distorted_radii
+  // calculation but saves calculating the `interpolation_info` twice.
+  const YlmSpherepack extended_ylm(l_max_ + 1, m_max_ + 1);
+  const auto interpolation_info =
+      extended_ylm.set_up_interpolation_info(theta_phis);
+
+  const DataVector coefs = functions_of_time.at(f_of_t_name_)->func(time)[0];
+  check_coefficients(coefs);
+  DataVector extended_coefs(extended_ylm.spectral_size(), 0.);
+
+  // Copy over the coefficients. The additional coefficients of order `l_max_
+  // +1` are zero and will only have an effect in the interpolation of the
+  // cartesian gradient.
+  SpherepackIterator extended_iter(l_max_ + 1, m_max_ + 1);
+  SpherepackIterator iter(l_max_, m_max_);
+  for (size_t l = 2; l <= l_max_; ++l) {
+    const int m_max = std::min(l, m_max_);
+    for (int m = -m_max; m <= m_max; ++m) {
+      iter.set(l, m);
+      extended_iter.set(l, m);
+      extended_coefs[extended_iter()] = coefs[iter()];
+    }
+  }
+
+  // Re-use allocation
+  auto& distorted_radii = get<0>(theta_phis);
+  extended_ylm.interpolate_from_coefs(make_not_null(&distorted_radii),
+                                      extended_coefs, interpolation_info);
+  // Calculates the Pfaffian derivative at the internal collocation points of
+  // YlmSpherePack. We can't interpolate these directly as they are not smooth
+  // across the poles, so we convert them to the Cartesian gradients first,
+  // which are smooth.
+  const auto angular_gradient =
+      extended_ylm.gradient_from_coefs(extended_coefs);
+
+  tnsr::i<DataVector, 3, Frame::Inertial> cartesian_gradient(
+      extended_ylm.physical_size());
+
+  // Re-use allocations
+  std::array<DataVector, 2> collocation_theta_phis{};
+  collocation_theta_phis[0].set_data_ref(&get<2>(cartesian_gradient));
+  collocation_theta_phis[1].set_data_ref(&get<1>(cartesian_gradient));
+  collocation_theta_phis = extended_ylm.theta_phi_points();
+
+  const auto& col_thetas = collocation_theta_phis[0];
+  const auto& col_phis = collocation_theta_phis[1];
+
+  // The Cartesian derivative is the Pfaffian derivative multiplied by the
+  // inverse Jacobian matrix. Some optimizations here may be possible by
+  // introducing temporaries for some of the sin/cos which are computed twice,
+  // if the compiler CSE doesn't take care of it.
+  get<0>(cartesian_gradient) =
+      (cos(col_thetas) * cos(col_phis) * get<0>(angular_gradient) -
+       sin(col_phis) * get<1>(angular_gradient));
+
+  get<1>(cartesian_gradient) =
+      (cos(col_thetas) * sin(col_phis) * get<0>(angular_gradient) +
+       cos(col_phis) * get<1>(angular_gradient));
+
+  get<2>(cartesian_gradient) = -sin(col_thetas) * get<0>(angular_gradient);
```

Comment:
This cartesian_gradient is already computed by the function DxRadius in ApparentHorizons/Tags.hpp. (here "Radius" is the thing being expanded in ylms, which for a Strahlkorper is the radius).  But DxRadius is a ComputeItem. :(

Seems to me that DxRadius in ApparentHorizons/Tags.hpp should be moved out of a ComputeItem and into a free function (and the ComputeItem DxRadius should then call the function), and then the code here should also call that same function. This might be beyond the scope of this PR however, so I would be fine with an issue.

Reply:
@markscheel I remember trying to use this but being confused as to what the `radius` there is and why all the results are multiplied with `one_over_r` which I did not need. I assume for a Strahlkorper this radius is just 1?

but yes, I think a major refactoring is necessary for `ApparentHorizons`, including:

- moving `YlmSpherepack` out of `ApparentHorizons`, so `CoordinateMaps` and possibly other libraries in the future do not depend on `ApparentHorizons`. Ideally done before the first cyclical dependency pops up.
- refactoring all major compute tags into free functions. Quite a few utility functions such as coordinate transformations between euclidean and spherical including Jacobian, Pfaffian, Hessian, etc. might also be needed in Spectre outside of `ApparentHorizons` - they certainly are in the shape map.

Perhaps this is already enough to justify a `SphereUtilities` library? This should avoid quite a bit of code duplication and encourage consistency across the libraries.

Reply:
I like the idea of a `SphereUtilities` library 👍

Reply:
Yes, I like the idea of `SphereUtilities` too.

In Strahlkorper, the radius is not 1; the radius is the thing being expanded in ylms.  That is, `radius(theta,phi)` defines the surface of a Strahlkorper, or equivalently `radius = radius_lm Y_lm` where `radius_lm` are the coefficients. Isn't that the same for the shape map? The reason that the Jacobians/Hessians factor out powers of 1/r is that r*Jacobian is independent of the actual shape of the surface and therefore can be computed only once, and similar for Hessians.

Reply:
hmm, I remember trying to use the `InvJacobianCompute` function followed by the `DxRadius` function to calculate the cartesian derivative. However, dividing by the radius in `DxRadius` spoiled things so I am not doing that now. I think there was some confusion when the radius was the radius of the sphere we are developing our scalar function on and when it was the actual scalar function we are developing. Either way, the finite difference test shows that this is the correct calculation of the jacobian so we should not need to worry for the shape map. 

I will start an issue about `SphereUtilities`.